mount: Put rclone mount in the background - fixes #723

ishuah opened the pull request:
This PR enables mount to run in the background. Windows is not supported at the moment (daemon package does not currently support windows.)

ishuah commented:
@ncw looks like the daemon package comes with several compile errors...

ncw commented on cmd/mountlib/mount.go:
```diff
@@ -174,6 +176,32 @@ will see all files and folders immediately in this mode.
 				}
 			}
 
+			// Windows is not supported for background tasks
```
In order to fix the compile problems (and it will be neater for windows) you need to factor this bit of code into a separate file and mark it with build tags for the OSes the daemon package will compile for.  This should export a single function background() say, which runs in background mode.

Then make a file which compiles for the opposite of the first file and that can just log that it isn't supported.

Does that make sense?

Check out the docs on [build tags](https://golang.org/pkg/go/build/#hdr-Build_Constraint) and see the example [syslog.go](https://github.com/ncw/rclone/blob/master/fs/log/syslog.go) and [syslog_unix.go](https://github.com/ncw/rclone/blob/master/fs/log/syslog_unix.go)

ishuah replied:
Got it!

ncw commented on docs/content/commands/rclone_mount.md:
```diff
@@ -20,6 +20,10 @@ This is **EXPERIMENTAL** - use with care.
 
+You can either run mount in foreground mode or a background mode. Default is background
+mode, use the `--foreground` flag to specify foreground mode. Background mode is only
+supported on Linux and OSX, you an only run it in foreground mode on Windows.
```
typo here `you an only`

ncw commented on cmd/mountlib/mount.go:
```diff
@@ -196,7 +224,7 @@ will see all files and folders immediately in this mode.
 	flags.StringArrayVarP(flagSet, &ExtraOptions, "option", "o", []string{}, "Option for libfuse/WinFsp. Repeat if required.")
 	flags.StringArrayVarP(flagSet, &ExtraFlags, "fuse-flag", "", []string{}, "Flags or arguments to be passed direct to libfuse/WinFsp. Repeat if required.")
-	//flags.BoolVarP(flagSet, &foreground, "foreground", "", foreground, "Do not detach.")
+	flags.BoolVarP(flagSet, &Foreground, "foreground", "", Foreground, "Do not detach.")
```
This changed the default to background... There are quite a lot of rclone mount users and changing the default might be unfriendly....  What do you think?

ishuah replied:
I thought about it while building this and I think it's a drastic change. Default foreground is much nicer, won't change how existing users interact with mount.

ncw replied:
I suggest we change the name of the flag to `--background` then too!

You might want to write in the docs also that if using `--background` you'll probably want to use `--log-file filename` or `--syslog` to capture the output of the mount.

seuffert replied:
"--daemon" like rsync? would be familiar for many

ishuah replied:
I like `--daemon` better

ishuah commented:
Amended my implementation to fix the Windows compile problems, works great!

Looks like `github.com/sevlyar/go-daemon` requires [`cgo` to successfully compile](https://github.com/sevlyar/go-daemon/blob/master/lock_file_darwin.go#L10) for `darwin`, that's why `make compile_all` is [failing](https://travis-ci.org/ncw/rclone/jobs/346955192).

I switched to using foreground mode as default, background(daemon) mode as an option. Changed the flag to `--daemon`.

ncw commented:
> Looks like github.com/sevlyar/go-daemon requires cgo to successfully compile for darwin, that's why make compile_all is failing.

I suggest you add an additional build constraint then to make the darwin stuff build with and without cgo.

So I think you want to build the null implementation for

(windows || (darwin && !cgo))

    // +build windows darwin,!cgo

A bit of boolean algebra to find the inverse

!(windows || (darwin && !cgo))
= !windows && !(darwin && !cgo)
= !windows && (!darwin || cgo)

Which is in build constraint language

    // +build !windows
    // +build !darwin cgo

I found the only way I can get the build constraints right is with boolean algebra - you may have a better intuitive feel for it!

Does that look right?

> I switched to using foreground mode as default, background(daemon) mode as an option. Changed the flag to --daemon.

perfect :-)

PS I found a bug in your 4c1ffc7f54919d40a566ec5531e236ca72bcc010 which was [reported in the forum](https://forum.rclone.org/t/rclone-v1-39-171-g1383df4f-has-crashed/4944/2) and by the time I had worked out what was happening I had made the fix in 7d4da1c66ad434d85452b311f6acd36ca8727037.  I award myself one :-1: for not noticing this in the code review!

ishuah commented:
> PS I found a bug in your 4c1ffc7 which was reported in the forum and by the time I had worked out what was happening I had made the fix in 7d4da1c. I award myself one 👎 for not noticing this in the code review!

I should have checked that error to begin with,  I'll be more careful!

I'll update the build constraints, that should fix the `darwin` issue.
Any idea about the AppVeyor test fail?

ishuah commented:
Got a chance to test the  build constraints, works perfectly! How will we ensure the final build includes `cgo` for darwin?

ncw commented:
That is looking good!

Can you rebase onto master to pick up e0eb666dbf415ad16826c647e9a986bb9722e76c just so we can see a clean test run?

ncw commented:
Looks good - please merge :-)
